yxi/xml/xml_generator.py

```diff
@@ -0,0 +1,113 @@
+import sys
+import json
+from xml.etree.ElementTree import Element, SubElement, tostring
+from xml.dom import minidom
+from math import log2
+
+"""
+This file takes in a `.yxi` description and outputs a xml suitable for a `kernel.xml` file
+can be used to package an xclbin for the Xilinx XRT runtime.
+See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+for the spec this is based on.
+"""
+size_key = "total_size"
+width_key = "data_width"
+
+def gen_xml(yxi):
+  mems = yxi["memories"]
+  check_mems_wellformed(mems)
+
+  root = Element("root", {"versionMajor": "1", "versionMinor": "6"})
+  kernel = SubElement(root, "kernel", {
+    "name" : yxi["toplevel"],
+    "language": "ip_c",
+    # TODO: Make sure this matches component.xml, Namely the `Toplevel` part.
+    #  See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+    "vlnv": "capra.cs.cornell.edu:kernel:Toplevel:1.0",
+    "attributes": "",
+    "preferredWorkGroupSizeMultiple" : "0",
+    "workGroupSize" : "1",
+    "hwControlProtocol": "ap_ctrl_hs"
+  })
+
+  # Construct ports
+  ports = SubElement(kernel, "ports")
+  # The subordinates XRT - AXI controller is added outside of the programs memory interface.
+  SubElement(ports, "port", {
+    "name": "S_AXI_CONTROL",
+    "mode": "slave",
+    "range": "0x1000", # XXX: Why is this 0x1000? Taken from the Xilinx examples  Maybe could be dynamically calculated?
+    "dataWidth" : "32",
+    "portType" : "addressable",
+    "base" : "0x0"
+  })
+
+  for mem in mems:
+    SubElement(ports, "port", {
+      "name": f"m_axi_{mem['name']}",
+      "mode": "master",
+      #NOTE(nathaniel): In the Xilinx examples range is usually 0xFFFFFF... but this should be fine for us?
+      "range": f"{hex(size_in_bytes(mem))}",
+      "dataWidth": f"{mem[width_key]}",
+      "portType": "addressable",
+      "base": "0x0"
+    })
+
+  # Construct Args
+  args = SubElement(kernel, "args")
+  # XRT spec starts args addresses at 0x10
+  args_addr = 0x10
+  for i, mem in enumerate(mems):
+    SubElement(args, "arg", {
+      "name": f"{mem['name']}",
+      "addressQualifier": "1", # 1 denotes the arguments as a global memory, 
+      "id" : f"{i}",
```

Comment:
I note that the old version uses `i+1`, FWIW:
https://github.com/calyxir/calyx/blob/b55a0626c44a1c2e185f0e300554a2c374900d40/calyx-backend/src/xilinx/xml.rs#L190

Reply:
The [latest XRT example](https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File) starts `id` at 0:

```
<args>
      <arg name="axi00_ptr0" addressQualifier="1" id="0" port="m00_axi" size="0x8" offset="0x010" type="int*" 
         hostOffset="0x0" hostSize="0x8"/> 
    </args>
```
So I think this is ok.

Reply:
Cool cool!